src/gui/torrentcontenttreeview.cpp

```diff
@@ -107,6 +108,78 @@ void TorrentContentTreeView::keyPressEvent(QKeyEvent *event)
     }
 }
 
+void TorrentContentTreeView::setupDownloadPriorityMenu(QMenu *menu, const bool createSubMenu) const
+{
+    const auto model = qobject_cast<TorrentContentFilterModel *>(this->model());
+    Q_ASSERT(model);
+
+    const auto getSelectedRows = [this]() { return selectionModel()->selectedRows(TorrentContentModelItem::COL_PRIO); };
+
+    const auto applyPriorities = [&](const BitTorrent::DownloadPriority priority)
+    {
+        return [model, getSelectedRows, priority]()
+        {
+            for (const QModelIndex &index : asConst(getSelectedRows()))
+            {
+                model->setData(index, static_cast<int>(priority));
+            }
+            emit model->filteredFilesChanged();
+        };
+    };
+    const auto applyPrioritiesByOrder = [model, getSelectedRows]()
+    {
+        // Equally distribute the selected items into groups and for each group assign
+        // a download priority that will apply to each item. The number of groups depends on how
+        // many "download priority" are available to be assigned
+
+        const QModelIndexList selectedRows = getSelectedRows();
+        const qsizetype priorityGroups = 3;
+        const auto priorityGroupSize = std::max<qsizetype>(selectedRows.length() / priorityGroups, 1);
+
+        for (qsizetype i = 0; i < selectedRows.length(); ++i)
+        {
+            auto priority = BitTorrent::DownloadPriority::Ignored;
+            switch (i / priorityGroupSize)
+            {
+            case 0:
+                priority = BitTorrent::DownloadPriority::Maximum;
+                break;
+            case 1:
+                priority = BitTorrent::DownloadPriority::High;
+                break;
+            default:
+            case 2:
+                priority = BitTorrent::DownloadPriority::Normal;
+                break;
+            }
+
+            model->setData(selectedRows[i], static_cast<int>(priority));
+        }
+
+        emit model->filteredFilesChanged();
```

Comment:
Looks incorrect to me. The model should emit its own signals by itself as an indication of some events/changes.

Reply:
Agree. I'm preparing another PR which changes that. To comprehend the big picture, first I must describe few things.

#### The "add new torrent" dialog
There is a bug. The dialog contains a label - disk space required by the torrent if downloaded:
![qb3](https://user-images.githubusercontent.com/68896601/141697778-28ce0805-8b97-4804-9b26-01dd79febc75.png)
The label gets updated whenever particular files contained by a torrent get enabled/disabled via checkboxes. However, it doesn't get updated if changing file priorities (i.e. from "normal" to "do not download"), either via right-click menu or via drop-down menu. It works for checkboxes thanks to this connection: https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/addnewtorrentdialog.cpp#L712 and this emit (last line): https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/torrentcontentmodel.cpp#L278-L300 No such facility exists for priorities.

#### The properties widget
Torrent properties widget handles check boxes in similar way. It connects to `TorrentContentFilterModel::filteredFilesChanged`: https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/properties/propertieswidget.cpp#L104 which is connected to  `TorrentContentModel::filteredFilesChanged` https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/torrentcontentfiltermodel.cpp#L37 While "add new torrent" dialog is updating a label, the properties widget uses the signal to send newly set priorities back to libtorrent: https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/properties/propertieswidget.cpp#L882-L886 In case of checkboxes, signal flows from inside the model. In other cases priorities are handled differently. We have 3 different mechanisms in total. For drop-down menu, signal flows from `PropListDelegate::setModelData` https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/properties/proplistdelegate.cpp#L113 via this connection: https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/properties/propertieswidget.cpp#L106 For the right-click menu, the signal isn't used at all, `applyPriorities` is called directly: https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/properties/propertieswidget.cpp#L658-L659

#### This PR
I moved handling of the right-click menu to `TorrentContentTreeView` so it can't call `applyPriorities` directly. I decided to use `TorrentContentFilterModel::filteredFilesChanged` since it's already connected to right place. I agree on that this solution is far from perfect. This could be resolved by doing something similar to what is done in `PropListDelegate` - add a new signal to `TorrentContentTreeView` class and make a connection chain: `QAction::triggered` -> `TorrentContentTreeView::filteredFilesChanged` -> `PropertiesWidget::filteredFilesChanged`. The disk space label bug could be fixed similarly by connecting `TorrentContentTreeView::filteredFilesChanged` -> `AddNewTorrentDialog::filteredFilesChanged`. But this is getting so complicated - new signal, new connections, still 3 different ways for handling file priorities.

#### New PR
The idea is to handle file priorities in one unified way - the way that is used for checkboxes currently. Basically it's about emitting signal from inside the model whenever priorities are edited - from inside `TorrentContentModel::setData`. Not only for `Qt::CheckStateRole`, but also for `Qt::EditRole`. This way the label bug will be fixed, `PropListDelegate::filteredFilesChanged` signal won't be needed at all, `TorrentContentTreeView` won't need to communicate neither with `PropertiesWidget` nor with `AddNewTorrentDialog`. There is only one obstacle. Simply emitting signal inside `TorrentContentModel::setData` is suboptimal. Priorities may be changed for multiple files in one go, in a loop, when using right-click menu while multiple items are selected. Signal would be emitted for every selected file while we only need it to be emitted once. This can be resolved by providing facility for bulk-changing file priorities. I figured out two ways of doing so.

#### First way - implicit.
It's about doing everything implicitly by the model. The model emits the signal in response to `setData` but not right away, it holds until new priorities stop flowing in. This is achieved by a 0-interval `QTimer`. Such timer ticks when QT event queue gets empty - after all posted events are processed. The timer can be stared multiple times but it will tick only once. Implementation is here: 6aa49ca0c400f758fa1f7ff40ea95c1ab100d56a

#### Second way - explicit.
This way is less universal. It's about providing dedicated function for bulk-changing file priorities. The model has a new method - `changeFilePriorities` for changing priorities of multiple files. The old way of changing priorities - `setData` - is still there and is used by checkboxes and drop-down menus. Right-click menus use `changeFilePriorities` explicitly. Implementation: 0cfae1afa45148227e208c4c7cafae621fbaa5c1

#### Yet another PR
Maybe bulk-changing file priorities is not needed at all. Perhaps simply emitting the signal every time `setData` is called is good enough. But my proposal lays foundations for yet another PR. Currently, when file priorities are being changed via checkboxes, entire torrent content tree view gets updated right away - all values in all columns including "progress" and "availability". This is different in case of changing priorities via right-click menu or via drop-down menu. Values get updated, but not right away, there is a "lag". The code for checkboxes is here: https://github.com/qbittorrent/qBittorrent/blob/b45248bf99ef00ce314d210c1cb9ebe5fe247cc0/src/gui/torrentcontentmodel.cpp#L297-L298 No such thing exists for menus. Values get updated after some time because of the periodic update via `PropertiesWidget::loadDynamicData`. My yet-another-proposal is to change that - always update all values in the tree view right away, not only when using checkboxes but also when using menus. Code for the two ways respectively is here: 13a2573f4ee96164df9478836a17d1e8985db046 cc0782913d5aac6617fed6d0fd10ce86c02e88d5. It not only removes the lag, but also unifies handling of editing priorities even more. Together with the bulk-changing of priorities it's optimal - priorities may be set on multiple files and entire treeview will be updated only once.

#### Epilogue
The decision on how to proceed is in your hands. Personally I would leave the "Looks incorrect to me" code just like it is, for now. There are ways to fix it - add new signal to `TorrentContentTreeView` and chain-connect it, simply emit `filteredFilesChanged` inside `TorrentContentModel::setData` also for `Qt::EditRole`, implement bulk-changing file priorities... It seems to me that all sensible ones need another PR.

Another idea that I didn't mention yet - pass callback function to `TorrentContentTreeView::setupPriorityMenu` that would be called on menu click. `PropertiesWidget` would pass `applyPriorities` and `AddNewTorrentDialog` would pass `updateDiskSpaceLabel`.